File: langs/ocaml/prjs/isses/M001.ml
open String

#use "../../utils/tester.ml"
#use "../../utils/from_file.ml"
#use "../../utils/print_util.ml"
#use "../../utils/to_string.ml"
#use "../../utils/to_wraper.ml"
#use "../../utils/from_wrap.ml"
#use "../../utils/from_string.ml"
#use "../../utils/from_tuple.ml"
#use "../../utils/for_paiza.ml"
#use "../../utils/util.ml"

module L = List
module S = String
module A = Array

(* ts => sI,sJoin  fs => to_i,to_nums  tw => num0to,replicate,lines,words,take,drop,any fw => isin,notin, maxs,mins,sums*)
let m001 = function f::s::[] ->  let first = words f and second = words s in first @ second | _ -> failwith "err"
(* let test() = let res = m001 ["fishing gardening swimming fishing";"hunting fishing fishing biting"] in pSs res *)
let rec uniqe = function [] -> [] | x::xs -> if L.exists (fun s -> s = x) xs then uniqe xs else x :: (uniqe xs)
(* let test() = let res = uniqe ["fishing";"gardening";"swimming";"fishing";"hunting";"fishing";"fishing";"biting"] in pSs res *)
let showMap ms = A.iter(fun (k,v) -> printf "(%s,%d)" k v ) ms
let initMap un = let un_a = A.of_list un in A.map(fun a -> (a,0)) un_a
(* let test() = let res = initMap ["hoge";"fuga";"mamo"] in showMap res *)
let setMap s map = for i = 0 to (A.length map)-1 do let (k,v) = map.(i) in if k = s then map.(i) <- (k,v+1) else map.(i) <- (k,v) done
(* let test() = let res = [|("mamo",0);("hoge",0);("fuga",0)|] in setMap "hoge" res; showMap res *)
let domain ss = let un = uniqe ss in let map = initMap un in L.iter (fun s -> setMap s map) ss; map
(* let test() = let res = domain ["fishing";"gardening";"swimming";"fishing";"hunting";"fishing";"fishing";"biting"] in showMap res *)
let domains ss = let fs = m001 ss in let map = domain fs in let ns = A.map(fun (k,v) -> v) map in
  let res = maxs(A.to_list ns) in [sI res]
let test() = let res = domains ["fishing gardening swimming fishing";"hunting fishing fishing biting"] in pSs res
(* main *)

let user() = let line = read_line() in let lim = (length line) - 1 in let lines = L.map(fun _ -> read_line())(replicate lim 0) in line ::lines

let product() = let u = user()
  in ps(sSL u)
  (* in let ans =domains u in List.iter(fun an -> ps an) ans *)

let developer exp ans name = zz_act ~dsp:name "M001 Test" (sSs ans) (sSs exp)

let develop() =
  let issue1 = "a:/pj/mamo/refs/isses/issue1.txt" and result1 = "a:/pj/mamo/refs/isses/result1.txt" in
  let issue2 = "a:/pj/mamo/refs/isses/issue2.txt" and result2 = "a:/pj/mamo/refs/isses/result2.txt" in
  let issue3 = "a:/pj/mamo/refs/isses/issue3.txt" and result3 = "a:/pj/mamo/refs/isses/result3.txt" in
  let issue4 = "a:/pj/mamo/refs/isses/issue4.txt" and result4 = "a:/pj/mamo/refs/isses/result4.txt" in
  let issue5 = "a:/pj/mamo/refs/isses/issue5.txt" and result5 = "a:/pj/mamo/refs/isses/result5.txt" in
  let issue6 = "a:/pj/mamo/refs/isses/issue6.txt" and result6 = "a:/pj/mamo/refs/isses/result6.txt" in
  developer (domains (ffUser issue1)) (ffUser result1) "test 1";
  developer (domains (ffUser issue2)) (ffUser result2) "test 2";
  developer (domains (ffUser issue3)) (ffUser result3) "test 3";
  developer (domains (ffUser issue4)) (ffUser result4) "test 4";
  developer (domains (ffUser issue5)) (ffUser result5) "test 5";
  developer (domains (ffUser issue6)) (ffUser result6) "test 6"

let refactor() = let issue = "a:/pj/mamo/refs/isses/issue1.txt" and result = "a:/pj/mamo/refs/isses/result1.txt" in
  let arg = ffUser(issue) and exp = ffUser(result)
  (* in ps(sSs arg); ps(sSs exp) *)
  in let ans = domains arg in zz_act ~dsp:"refactor" "M001 Test" (sSs ans) (sSs exp)

(* main M001 m001 domain user *)
let() =
  (* test() *)
  (* refactor() *)
  develop()
  (* product() *)

File: langs/c/prjs/isses/M001.c
#include "../../utils/sizes.h"
#include "../../utils/to_string.c"
#include "../../utils/from_file.c" 
#include "../../utils/tester.c"
#include "../../utils/from_string.c"
#include "../../utils/from_heap.c"
#include "../../utils/print_util.c"
#include "../../utils/to_wrap.c"
#include "../../utils/from_wrap.c"

static int arg_sz, ans_sz;

void showSs(char ss[][WORD], int size){
  char chars[LINE] = {0}; chars[0] = '[';
  for(int i= 0; i<size-1; i++){ strcat(chars,ss[i]); strcat(chars,",");}
  strcat(chars,ss[size-1]); strcat(chars,"]"); ps(chars);
}
// void test(){ char arg[][WORD] = {"hoge","fuga","moge"}; showSs(arg,3);}
void m001(char *str, char dmt, char res[][WORD], int size){
  int id = 0, id_str = 0;
  for(int i= 0; i<size; i++){
    for(int j= 0; j<WORD; j++){
      char ch = str[id_str];
      if(ch == dmt || ch == '\0'){
        if(ch == dmt){ res[i][id] = '\0'; id = 0; id_str++; break;}
        if(ch == '\0'){ res[i][id] = '\0'; return;}
      }
      res[i][id] = str[id_str]; id++; id_str++;
    }
  }
}
// void test(){ char str[] = {"hoge fuga mamo"}; char res[3][WORD]; m001(str, ' ', res, 3); showSs(res,3);}
int domain(char all[][WORD], int size){
  int res[size];
  for(int i= 0; i<size; i++){
    int count = 0;
    for(int j= 0; j<size; j++){
      if(strcmp(all[i], all[j]) == 0){ count++;}
    }
    res[i] = count;
  }
  return maxs(res,size);
}
// void test(){ char arg[][WORD] = {"hoge","fuga","mamo","fuga"}; int res = domain(arg,4); pi(res);}
// main develop
char **domains(char **arg, int sz){
  arg_sz = sz; ans_sz = 1; int size = 1;
  for(int i= 0; i<(int)strlen(arg[0]); i++){ if(arg[0][i] == ' '){ size++;}}
  char fst[size][WORD]; m001(arg[0], ' ', fst, size);
  char snd[size][WORD]; m001(arg[1], ' ', snd, size);
  char all[size*2][WORD]; int id = 0;
  for(int i= 0; i<size; i++){ strcpy(all[id++],fst[i]);}
  for(int i= 0; i<size; i++){ strcpy(all[id++],snd[i]);}
  int res = domain(all,size*2);
  char **ans = fhSs(ans_sz);
  ans[0] = sI(res);
  return ans;
}
void test(){
  char ** args = fhSs(2); 
  args[0] = fh_line("fishing gardening swimming fishing");
  args[1] = fh_line("hunting fishing fishing biting");
  char **res = domains(args,2); pSs(res,ans_sz);
  // develop();
}  // exp is exp

UserFF user() {
  char line[LINE]; 
  fgets(line, LINE, stdin);
  int lim = strlen(line) - 1;
  char **res = fhSs(lim);
  res[0] = fh_line(line);
  for(int i= 1; i<lim; i++){ fgets(line, LINE, stdin); res[i] = fh_line(trim(line));}
  return ff_set(lim, res);
}
void product() {
  UserFF u = user();
  // pSs(u.lines, u.size);
  char **res = domains(u.lines,u.size); for(int i= 0; i<ans_sz; i++){ ps(res[i]);}
}

void developer(char *issue, char *result, char *name){
  UserFF arg = ff_list(issue), exp = ff_list(result); char **ans = domains(arg.lines,arg.size);
  zz_act("M001 Test", sSs(ans,ans_sz), sSs(exp.lines,ans_sz), name);
  fhSsFree(exp.lines,exp.size); fhSsFree(ans,ans_sz); fhSsFree(arg.lines,arg.size);
}
void develop() {
  char *issue1 = "a:/pj/mamo/refs/isses/issue1.txt", *result1 = "a:/pj/mamo/refs/isses/result1.txt",
       *issue2 = "a:/pj/mamo/refs/isses/issue2.txt", *result2 = "a:/pj/mamo/refs/isses/result2.txt",
       *issue3 = "a:/pj/mamo/refs/isses/issue3.txt", *result3 = "a:/pj/mamo/refs/isses/result3.txt",
       *issue4 = "a:/pj/mamo/refs/isses/issue4.txt", *result4 = "a:/pj/mamo/refs/isses/result4.txt",
       *issue5 = "a:/pj/mamo/refs/isses/issue5.txt", *result5 = "a:/pj/mamo/refs/isses/result5.txt",
       *issue6 = "a:/pj/mamo/refs/isses/issue6.txt", *result6 = "a:/pj/mamo/refs/isses/result6.txt";
  developer(issue1, result1,"test-1"); developer(issue2, result2,"test-2");
  developer(issue3, result3,"test-3"); developer(issue4, result4,"test-4");
  developer(issue5, result5,"test-5"); developer(issue6, result6,"test-6");
}

void refactor() {
  char *issue = "a:/pj/mamo/refs/isses/issue1.txt", *result = "a:/pj/mamo/refs/isses/result1.txt";
  UserFF arg = ff_list(issue), exp = ff_list(result); 
  // ps(ff_sSs(arg)); ps(ff_sSs(exp)); pi(arg.size);
  char **ans = domains(arg.lines,arg.size);
  zz_act("M001 Test", sSs(ans,ans_sz),  sSs(exp.lines,ans_sz),"refactor");
}

// M001 m001 domain user
int main(void){
  // test();
  // refactor();
  develop();
  // product();
}